File: Graphs/DirectedAcyclicGraphs/source/directed_acyclic_graph.py
from collections import defaultdict, deque


class DirectedAcyclicGraph:
    def __init__(self):
        self.graph = defaultdict(list)
        self.vertices = set()

    def add_vertex(self, vertex):
        self.vertices.add(vertex)

    def add_edge(self, source, target):
        if source not in self.vertices:
            self.add_vertex(source)
        if target not in self.vertices:
            self.add_vertex(target)

        self.graph[source].append(target)
        if self._has_cycle():
            self.graph[source].pop()
            raise ValueError(f"Adding edge {source} -> {target} introduces a cycle!")

    def dfs_iterative(self, start_vertex, process=None, record_stack=None, detect_cycle=False):
        visited = set()
        rec_stack = set()
        stack = [(start_vertex, 0)]

        while stack:
            vertex, neighbor_index = stack[-1]

            if vertex not in visited:
                visited.add(vertex)
                if detect_cycle:
                    rec_stack.add(vertex)
                if process:
                    process(vertex)

            neighbors = self.graph[vertex]

            if neighbor_index >= len(neighbors):
                if record_stack:
                    record_stack.appendleft(vertex)
                if detect_cycle:
                    rec_stack.remove(vertex)
                stack.pop()
                continue

            neighbor = neighbors[neighbor_index]
            stack[-1] = (vertex, neighbor_index + 1)

            if detect_cycle and neighbor in rec_stack:
                return True

            if neighbor not in visited:
                stack.append((neighbor, 0))

        if detect_cycle:
            return False

    def _has_cycle(self):
        for vertex in self.vertices:
            if self.dfs_iterative(vertex, detect_cycle=True):
                return True
        return False

    def topological_sort(self):
        visited = set()
        stack = deque()

        for vertex in self.vertices:
            if vertex not in visited:
                self.dfs_iterative(vertex, record_stack=stack)

        return list(stack)

    def display(self):
        for vertex in self.vertices:
            print(f"{vertex} -> {self.graph[vertex]}")

File: Graphs/DirectedAcyclicGraphs/source/directed_acyclic_graph.js
class DirectedAcyclicGraph {
    constructor() {
        this.graph = {};
        this.vertices = new Set();
    }

    addVertex(vertex) {
        this.vertices.add(vertex);
        if (!this.graph[vertex]) {
            this.graph[vertex] = [];  // Initialize an empty array for the vertex
        }
    }

    addEdge(source, target) {
        if (!this.vertices.has(source)) {
            this.addVertex(source);
        }
        if (!this.vertices.has(target)) {
            this.addVertex(target);
        }

        if (!this.graph[source]) {
            this.graph[source] = [];
        }

        this.graph[source].push(target);
        if (this._hasCycle()) {
            this.graph[source].pop();
            throw Error(`Adding edge ${source} -> ${target} introduces a cycle!`);
        }
    }

    dfsIterative(start_vertex, visited = null, process = null, record_stack = null, detect_cycle = false) {
        if (!visited) {
            visited = new Set();
        }
        const rec_stack = new Set();
        const stack = [start_vertex];

        while (stack.length > 0) {
            const vertex = stack.pop();

            if (!visited.has(vertex)) {
                visited.add(vertex);
                if (detect_cycle) {
                    rec_stack.add(vertex);
                }

                if (process) {
                    process(vertex);
                }

                if (record_stack !== null) {
                    record_stack.unshift(vertex);
                }

                for (const neighbor of (this.graph[vertex] || []).reverse()) {
                    if (detect_cycle && rec_stack.has(neighbor)) {
                        return true;
                    }
                    if (!visited.has(neighbor)) {
                        stack.push(neighbor);
                    }
                }

                if (detect_cycle) {
                    rec_stack.delete(vertex);
                }
            }
        }

        if (detect_cycle) {
            return false;
        }
    }

    _hasCycle() {
        const visited = new Set();
        for (const vertex of this.vertices) {
            if (!visited.has(vertex)) {
                if (this.dfsIterative(vertex, visited, null, null, true)) {
                    return true;
                }
            }
        }
        return false;
    }

    display() {
        for (const vertex of this.vertices) {
            console.log(`${vertex} -> ${this.graph[vertex]}`);
        }
    }
}

module.exports = DirectedAcyclicGraph;
